modules/mapbox/src/deck-utils.ts

```diff
@@ -140,6 +155,53 @@ export function getViewState(map: Map): MapViewState & {
     padding: map.getPadding(),
     repeat: map.getRenderWorldCopies()
   };
+
+  if (map.getTerrain?.()) {
+    // When the base map has terrain, we need to target the camera at the terrain surface
+    centerCameraOnTerrain(map, viewState);
+  }
+
+  return viewState;
+}
+
+function centerCameraOnTerrain(map: Map, viewState: MapViewState) {
+  if (map.getFreeCameraOptions) {
+    // mapbox-gl v2
+    const {position} = map.getFreeCameraOptions();
+    if (!position || position.z === undefined) {
+      return;
+    }
+
+    // @ts-ignore transform is not typed
+    const height = map.transform.height;
+    const {longitude, latitude, pitch} = viewState;
+
+    // Convert mapbox mercator coordinate to deck common space
+    const cameraX = position.x * TILE_SIZE;
+    const cameraY = (1 - position.y) * TILE_SIZE;
+    const cameraZ = position.z * TILE_SIZE;
+
+    // Mapbox manipulates zoom in terrain mode, see discussion here: https://github.com/mapbox/mapbox-gl-js/issues/12040
+    const center = lngLatToWorld([longitude, latitude]);
+    const dx = cameraX - center[0];
+    const dy = cameraY - center[1];
+    const cameraToCenterDistanceGround = Math.sqrt(dx * dx + dy * dy);
+
+    const pitchRadians = pitch! * DEGREES_TO_RADIANS;
+    const altitudePixels = 1.5 * height;
+    const scale = (altitudePixels * Math.sin(pitchRadians)) / cameraToCenterDistanceGround;
+    viewState.zoom = Math.log2(scale);
+
+    const cameraZFromSurface = (altitudePixels * Math.cos(pitchRadians)) / scale;
+    const surfaceElevation = cameraZ - cameraZFromSurface;
+    viewState.position = [0, 0, surfaceElevation / unitsPerMeter(latitude)];
```

Comment:
Just trying to get confirmation on my understanding:
So, Deck's MapView will now have support for offsetting camera position from it's geolocation anchor point through `position`.

We're accessing mapbox free camera to calculate a scale value used to determine viewState `position.z` and `zoom` so that the camera can be moved as it traverses over terrain.

If that's the case, that's pretty great. The new position property is useful beyond terrain for viewing a variety of 3D datasets.

Reply:
Position should work in production already.